### DISCUSSION TITLE
ZFS built into kernel - mount bootfs

### DISCUSSION THREAD

**Discussion by furbyhaxx** [openzfs/zfs#15865] [Q&A]
Hi everybody,
If ZFS 2.2.2 is built into the kernel, is it able to mount the dataset with the bootfs flag automatically to root /?
Or do I still need an initrd?
I only found this discussion with the associated issue [https://github.com/openzfs/zfs/discussions/4300](https://github.com/openzfs/zfs/discussions/4300) which seems to be still not implemented.

**Comment by IvanVolosyuk** [comment]
I have ZFS compiled into kernel. I wasn't been able to boot kernel without
initrd. It contains zpool and zfs command line utilities for mounting and
it seems root=ZFS=zpool/ROOT is not recognised without initrd. It seems ZFS
relies on udev to find all relevant devices as well as the cache file.

**Reply by rincebrain** [reply]
root=ZFS is parsed out in the initrd, I believe, yes, so that wouldn't help you here.

Conceivably you could have some zfs.import_on_boot= module parameter that triggers the "import pools on module init" logic for a specific pool name or GUID or w/e, I imagine, but I don't actually know enough about how the dataset mount process works to guess if there's something in userspace it needs for that to be sufficient on its own...

Someone could try implementing it and see.

(The other problem becomes that then you might need a lot more than just "import pool X at this location"  to handle things like the systemd mount dependency parameters and all the fun that you usually use those and/or fstab for...but in the kernel.)
